qiskit/qpy/type_keys.py

```diff
@@ -135,6 +138,10 @@ def assign(cls, obj):
             return cls.NULL
         if obj is CASE_DEFAULT:
             return cls.CASE_DEFAULT
+        if isinstance(obj, list):
+            return cls.LIST
+        if isinstance(obj, dict):
+            return cls.DICT
```

Comment:
I have not dug deeply into the internals of qpy. Why do we need these `LIST` and `DICT` types? Why can we not just use `SEQUENCE` and `MAPPING`? (I see that `MAPPING` did not support nesting and  you extended it with `DICT` but could it not be extended with `DICT`?).

Reply:
I agree. @junnaka51 is there any reason not to use existing types?

Reply:
I cannot find `SEQUENCE` and `MAPPING` as types in https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/qpy/type_keys.py . 

If I understand correctly, we need an identifier of `LIST` (or we can call it `SEQUENCE`) and that of `DICT` (or `Mapping`) for reading them out from a binary. 
I now realize that we can use the existing functions such as
https://github.com/Qiskit/qiskit-terra/blob/e9f8b7c50968501e019d0cb426676ac606eb5a10/qiskit/qpy/common.py#L203
 instead of upgrading `qpy.binary_io.value.dumps_value()` and `qpy.binary_io.value.loads_value()`. But I think we still need the identifiers. For example, when we want to deal with `LIST` as a value in a `DICT`, then we may update `qpy.common.write_mapping()` as
```diff
def write_mapping(file_obj, mapping, serializer, **kwargs):
    """Write a mapping in the file like object.
    """
    num_elements = len(mapping)

    file_obj.write(struct.pack(formats.SEQUENCE_PACK, num_elements))
    for key, datum in mapping.items():
        key_bytes = key.encode(ENCODE)

+        if isinstance(datum, list):
+            type_key, datum_bytes = b"l", sequence_to_binary(datum, serializer, **kwargs)
+        else:
+            type_key, datum_bytes = serializer(datum, **kwargs)

        item_header = struct.pack(formats.MAP_ITEM_PACK, len(key_bytes), type_key, len(datum_bytes))
        file_obj.write(item_header)
        file_obj.write(key_bytes)
        file_obj.write(datum_bytes)
```
and `qpy.common.read_mapping()` as
```diff
def read_mapping(file_obj, deserializer, **kwargs):
    """Read a mapping from the file like object.
    """
    mapping = {}

    data = formats.SEQUENCE._make(
        struct.unpack(formats.SEQUENCE_PACK, file_obj.read(formats.SEQUENCE_SIZE))
    )
    for _ in range(data.num_elements):
        map_header = formats.MAP_ITEM._make(
            struct.unpack(formats.MAP_ITEM_PACK, file_obj.read(formats.MAP_ITEM_SIZE))
        )
        key = file_obj.read(map_header.key_size).decode(ENCODE)

+        if map_header.type == b"l":
+            datum = common.read_sequence(file_obj, deserializer, **kwargs)
+        else:
+             datum = deserializer(map_header.type, file_obj.read(map_header.size), **kwargs)
-        datum = deserializer(map_header.type, file_obj.read(map_header.size), **kwargs)
        mapping[key] = datum

    return mapping
```
, in both of which `b"l"` is used as the identifier of `LIST` (or `SEQUENCE` if we want). 
Of course this `b"l"` must not be defined in `qpy/type_keys.py`. But I thought `qpy/type_keys.py` summarizes all the types, so I put it there.

Reply:
You're absolutely right. I think nested dictionary is not a `Value` type, and indeed I didn't assume this data structure when I wrote this function. Instead of calling `value.dumps_value`, you can write a custom serializer function for discriminator/kernel config, which would look like

```python
def my_custom_serializer(obj, version):
    if isinstance(obj, dictionary):
	 with io.BytesIO() as container: 
            for key, value in obj.items()
                # recursively generate binary data
            binary_data = container.getvalue() 
        return b"D", binary_data
    return dumps_value(obj, version)
```

what do you think?

Reply:
@nkanazawa1989  Thank you so much for your suggestion 🙇 
Please give me a little more time. It has been taking time to understand it.

Reply:
Thanks, @junnaka51! By the way, this PR is currently labeled with the 0.25.0 milestone which should be released in about 10 days. There is no pressure to meet the milestone though. We can bump it back to the next milestone if needed. Leaving the milestone on the PR for now means that we can still merge it if it is ready (a newly opened PR would not generally by eligible for 0.25.0 at this point).

Reply:
@wshanks Thank you for the info! :)

Reply:
In any case, I think with your last push this PR might be good now? The one thing I wonder about is the lack of `type_keys` for `b"D"` and `b"l"` -- it feels like names would be clearer than plain strings, but I defer to @nkanazawa1989

Reply:
No, not yet. Although it works, I suppose how to deal with a nested dictionary is not yet well written.
I am trying to grasp @nkanazawa1989 `s comments.

Reply:
@nkanazawa1989 Thank you again for your advice and suggestion 🙇 
I agree that your approach is more sophisticated. I added a new serializer and deserializer comparable in abstraction to `value.dumps_value` and `value.loads_value`, respectively, and put `write_mapping` and `read_mapping` back to their original 👉 https://github.com/Qiskit/qiskit-terra/pull/10327/commits/cc878e657bffb5a671c07637de86322dbf537eae